File: 2024/golang/src/day2/main.go
package main

import (
	"math"
	"slices"
	"strconv"
	"utils"
)

func GetData(filename string) interface{} {
	lines := utils.GetLines(filename)
	var nums [][]int
	for _, line := range lines {
		nums = append(nums, utils.GetIntsFromLine(line, " "))
	}
	return nums
}

func PartOne(d interface{}) string {
	data := d.([][]int)
	s := 0
	for _, x := range data {
		if TrySafetyCheck(x) {
			s += 1
		}
	}
	return strconv.Itoa(s)
}

func PartTwo(d interface{}) string {
	data := d.([][]int)
	s := 0
	for _, x := range data {
		if TrySafetyCheck(x) {
			s += 1
		} else {
			for i := 0; i < len(x); i++ {
				copied := make([]int, len(x))
				copy(copied, x)
				if TrySafetyCheck(append(copied[:i], copied[i+1:]...)) {
					s += 1
					break
				}
			}
		}
	}
	return strconv.Itoa(s)
}

func TrySafetyCheck(x []int) bool {
	reversed := make([]int, len(x))
	copy(reversed, x)
	slices.Reverse(reversed)
	if slices.IsSorted(x) || slices.IsSorted(reversed) {
		for i := 0; i < len(x)-1; i++ {
			if math.Abs(float64(x[i]-x[i+1])) <= 0 || math.Abs(float64(x[i]-x[i+1])) >= 4 {
				return false
			}
		}
	} else {
		return false
	}
	return true
}

func main() {
	data := GetData("input.txt")
	utils.RunWithTimer(PartOne, data)
	utils.RunWithTimer(PartTwo, data)
}

File: 2024/python/day2/main.py
from utils.timers import run_with_timer, get_data_with_timer
from utils.input import read_input_as_int_arrays


def get_data(filename):
	return read_input_as_int_arrays(filename)


def part_one(d):
	return sum(1 for x in d if try_safety_check(x))


def part_two(d):
	c = 0
	for x in d:
		if try_safety_check(x):
			c += 1
		else:
			for i in range(len(x)):
				if try_safety_check(x[:i] + x[i+1:]):
					c += 1
					break
	return c


def try_safety_check(x):
	for i in range(len(x) - 1):
		sx = sorted(x)
		if (list(sx) != x and list(reversed(sx)) != x) or abs(x[i] - x[i+1]) not in [1, 2, 3]:
			return False
	return True


def main(f="input.txt"):
	data = get_data_with_timer(get_data, f)
	run_with_timer(part_one, data)
	run_with_timer(part_two, data)


if __name__ == '__main__':
	main()
